### ISSUE TITLE
Intellisense suggest nuint rather than null

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#68841]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Intellisense-suggest-nuint-rather-than-n/10404820)._

---
[severity:It's more difficult to complete my work]
When typing "nu":

![nuint.png](https://aka.ms/dc/image?name=B07cec8de2ed24c61ae3926fc021a8905638237064836952945_nuint.png&tid=07cec8de2ed24c61ae3926fc021a8905638237064836952945)

I want to complete to "null" but "nuint" has higher proprity and is always suggested first.

In my work null is more used than nuint and therefore null should be the first suggestion.

---
#### Feedback Bot on 6/30/2023, 02:28 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by sharwell** [comment]
Note that this only affects the first use. After `null` or `nuint` is used, it will continue to be in the MRU and be suggested at higher priority.

This seems like a case where target-typing could be taken into account.

**Comment by breakerofbuilds** [comment]
Name has type `string?`.

https://github.com/dotnet/roslyn/assets/104224536/342344e0-3a3a-4a70-9317-28ac758dc18e

**Comment by genlu** [comment]
Hmm, we made a targeted fix for this exact scenario a few years back. See: https://github.com/dotnet/roslyn/pull/44451
I'd need you to confirm the preference of `nunit` over `null` persists after restarting VS, to make sure it's not caused by MRU cache we use in VS completion.

@breakerofbuilds the video you shared looks like a razor file? If so, then it's an issue with VS LSP client, which has full control of item selection. I believe we'd have the same issue for our C# extension in vscode.

**Comment by davidwengier** [comment]
Yeah, I can repro this in Razor in 17.7 P6.
<img width="317" alt="image" src="https://github.com/dotnet/roslyn/assets/754264/263e4932-6b91-4954-a9f0-398fdf1856ec">

**Comment by richardhauer** [comment]
VS 17.12 Preview 2 and this is still a "problem". I put that in scare-quotes because the tool is working fine but it's a major source of frustration personally.

My `fat fingers` hit nuint at least 20 times per day, getting it permanently stuck in MRU. Sometimes I'll write it, delete and then write it again. It's hideously annoying. In nearly 35 years of programming I have never even once needed the `nuint` type in C# - I feel like this is an edge case for `unsafe` programming.

I know it's definitely possible to mark methods/properties/types as hidden from Intellisense. I feel like this keyword is a good candidate for that too. Given that most programmers will write `null` 500x per day and `nuint` perhaps once in a decade.

Better yet, some option where you can just manually hide some keyword suggestions from Intellisense so that people who do use this type (assuming they exist at all) won't be penalised.

**Comment by CyrusNajmabadi** [comment]
@richardhauer we'd take a community contribution here. Thanks!

**Comment by richardhauer** [comment]
Thanks for trying this out @CyrusNajmabadi 

I'm not across the detail, but I suspect the values are being driven by an MRU and that suggestions are increasingly influenced by Copilot. This might help to explain why we get different results.

The suggestion you are showing includes a full statement completion. I don't get that at all as I do not have Copilot enabled (we use Azure DevOps and not Github).

I looked at the fix you linked. I'm on 17.13.P1 and this issue still comes up for me quite a lot, so I don't really have an explanation outside of the dynamic MRU one. If I can avoid accidentally hitting the `nuint` option a few times in a row it seems to go back to suggesting `null`, which I think adds weight to my dynamic-MRU theory.

**Comment by CyrusNajmabadi** [comment]
The completion list is driven entirely by Roslyn here. The mru only comes into play *if* you select something else.

**Comment by richardhauer** [comment]
So if you select `nuint` by accident, even once, you just keep getting it?

That's basically what I'm saying. I haven't looked at the code so I don't know what it's supposed to be doing, I can only tell you what's happening to me (and others).

**Comment by CyrusNajmabadi** [comment]
> So if you select nuint by accident, even once, you just keep getting it?

It's an MRU.  So if you write 'null' again it will be more recent.  Presumably you're picking this again and again to put it back at the front of hte list (even once you pick `null`).  Given that we're not hearing about this from other people, and the code explicitly goes out of its way to select `null` over `nuint` until you explicitly override that, it seems like this is working as intended.

**Comment by richardhauer** [comment]
Yes - I acknowledged at the start that this was working as intended. See my original comment from Sep 22.

This isn't a bug per se. The suggestion is that it be _hidden altogether_, as is the case with other intellisense items (via `EditorBrowsableAttribute` for example, not saying that mechanism would work here, just that there's precedent for hiding rare use items from intellisense), since the ratio of `null`:`nuint` usage across user code must be about 1,000,000,000,000:1.

I'm glad someone put in some effort to de-prioritise `nuint` but it doesn't seem to be enough to stop people from tripping over it. Clearly, I'm not the only one as I didn't even raise the ticket. Further, both @davidwengier and @ghost above also shared a similar experience, so you clearly **are** hearing this from other people.

It's not such a heinous problem that most people would take the time to provide feedback, but it's still a fruit fly buzzing about that should be pretty easy to swat. It's one of those things that you don't really notice how annoying it is until it's gone.
